src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java

```diff
@@ -1,20 +1,28 @@
 package com.es.phoneshop.model.product;
 
-import java.math.BigDecimal;
-import java.util.ArrayList;
-import java.util.Currency;
-import java.util.List;
-import java.util.NoSuchElementException;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 
 public class ArrayListProductDao implements ProductDao {
+    private static ProductDao instance;
+
     private List<Product> products;
     private long maxId;
-    private final Object lock = new Object();
+    private static final Object lock = new Object();
 
-    public ArrayListProductDao(){
+    public static ProductDao getInstance() {
+        synchronized (lock) {
```

Comment:
Not optimal construction. Better to leave lock not static to use it in methods below, but in getInstance method use synchronization on class.